### PR TITLE
NIAD-509 sanbox environment

### PR DESCRIPTION
**Pull request by dawidskainos**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: docker/service/src/integration-test/java/uk/nhs/adaptors/scr/uat/common/TestData.java [hunk 1]
```diff
@@ -31,14 +31,15 @@ public static TestData build(List<Resource> resources) {
...
 
-        var fhirResponse = list.stream()
+        list.stream()
             .filter(resource -> resource.getFilename() != null)
             .filter(resource -> resource.getFilename().endsWith(RESPONSE_FILE_ENDING))
             .findFirst()
-            .orElseThrow(() -> new IllegalStateException(String.format("Missing %s resource", RESPONSE_FILE_ENDING)));
+            .ifPresent(it -> {
+                testDataBuilder.fhirResponse(TestData.readFile(it));
+                list.remove(it);
```
Comment (bartek-sarul) [inline docker/service/src/integration-test/java/uk/nhs/adaptors/scr/uat/common/TestData.java:40]: Will this work? Removing item while iterating?

File: docker/service/src/main/resources/application.yml [hunk 2]
```diff
@@ -1,7 +1,16 @@
 server:
-    port: ${SCR_SERVER_PORT:9000}
-    tomcat:
-        relaxed-query-chars: "|"
+  port: ${SCR_SERVER_PORT:9000}
+  tomcat:
+    relaxed-query-chars: "|"
```
Comment (bartek-sarul) [inline docker/service/src/main/resources/application.yml:4]: Missing from sandbox cfg

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: docker/service/src/integration-test/java/uk/nhs/adaptors/scr/uat/common/TestData.java (hunk 1 above)
Comment (dawidskainos) [reply docker/service/src/integration-test/java/uk/nhs/adaptors/scr/uat/common/TestData.java:40]: yes, works

File: docker/service/src/main/resources/application.yml (hunk 2 above)
Comment (dawidskainos) [reply docker/service/src/main/resources/application.yml:4]: not valid anymore - using the same codebase for prod and sandbox

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: docker/service/src/integration-test/resources/application.yml [hunk 3]
```diff
@@ -6,7 +11,14 @@ management:
...
         include: info, health, metrics
 
+spring:
+  mvc:
+    throw-exception-if-no-handler-found: true
+  resources:
+    add-mappings: false
+
 scr:
+  spineMode: ${SCR_SPINE_MODE:REAL}
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/integration-test/resources/application.yml]: Ok, changed.

File: docker/service/src/main/java/uk/nhs/adaptors/scr/clients/MockSpineClient.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Override
+    public ProcessingResult getScrProcessingResult(String contentLocation, long initialWaitTime, String nhsdAsid) {
+        return null;
+    }
+
+    @Override
+    public Response sendGetScrId(String requestBody, String nhsdAsid) {
+        String responseBody = getResourceAsString(EVENT_LIST_QUERY_RESPONSE);
+        return new Response(OK.value(), null, responseBody);
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/clients/MockSpineClient.java]: This story is only about creating Sandbox app. Any more sophisticated stubs will be created in https://gpitbjss.atlassian.net/browse/NIAD-703

File: docker/service/src/main/java/uk/nhs/adaptors/scr/clients/MockSpineClient.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.SneakyThrows;
+import org.apache.commons.io.IOUtils;
+import org.springframework.core.io.ClassPathResource;
+import uk.nhs.adaptors.scr.clients.SpineHttpClient.Response;
+import uk.nhs.adaptors.scr.models.ProcessingResult;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.springframework.http.HttpStatus.OK;
+
+public class MockSpineClient implements SpineClient {
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/clients/MockSpineClient.java]: Ok, changed

File: docker/service/src/main/java/uk/nhs/adaptors/scr/clients/ProdSpineClient.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import static org.springframework.http.HttpStatus.ACCEPTED;
+import static org.springframework.http.HttpStatus.OK;
+import static org.springframework.http.MediaType.TEXT_XML_VALUE;
+import static uk.nhs.adaptors.scr.consts.HttpHeaders.CONTENT_TYPE;
+import static uk.nhs.adaptors.scr.consts.HttpHeaders.NHSD_ASID;
+import static uk.nhs.adaptors.scr.consts.HttpHeaders.SOAP_ACTION;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Slf4j
+public class ProdSpineClient implements SpineClient {
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/clients/ProdSpineClient.java]: Ok, changed

File: docker/service/src/main/java/uk/nhs/adaptors/scr/clients/SpineClient.java [hunk 7]
```diff
@@ -1,157 +1,11 @@
...
-            }
-        }
-    }
-
-    @RequiredArgsConstructor
-    @Getter
-    public static class ScrRetryBackOffContext implements BackOffContext {
-        private final RetryContext retryContext;
-    }
+public interface SpineClient {
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/clients/SpineClient.java]: Ok, changed

File: docker/service/src/main/java/uk/nhs/adaptors/scr/components/OperationOutcomeExceptionHandler.java [hunk 8]
```diff
@@ -5,88 +5,127 @@
...
 
-    private MediaType getRequestMediaType(WebRequest request) {
-        String contentType = request.getHeader(HttpHeaders.CONTENT_TYPE);
-        if (contentType == null) {
-            throw new IllegalStateException("Request content type is null");
-        }
-        return MediaType.parseMediaType(contentType);
+    private ResponseEntity<Object> errorResponse(Exception ex, HttpHeaders headers, HttpStatus status,
+                                                 OperationOutcome operationOutcome) {
+        LOGGER.error("Creating OperationOutcome response for " + ex.getClass(), ex);
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/components/OperationOutcomeExceptionHandler.java]: removed

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ScrSpineMode.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package uk.nhs.adaptors.scr.config;
+
+public enum ScrSpineMode {
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/config/ScrSpineMode.java]: Ok, removed

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ScrSpringConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+@Configuration
+public class ScrSpringConfiguration {
+
+    @Autowired
+    private ScrConfiguration scrConfiguration;
+
+    @Bean
+    @Autowired
+    public SpineClient spineClient(SpineConfiguration spineConfiguration, SpineHttpClient spineHttpClient) {
+        switch (scrConfiguration.getSpineMode()) {
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/config/ScrSpringConfiguration.java]: Ok, changed to Boolean

File: docker/service/src/main/java/uk/nhs/adaptors/scr/components/OperationOutcomeExceptionHandler.java [hunk 11]
```diff
@@ -5,88 +5,127 @@
...
+import static uk.nhs.adaptors.scr.utils.OperationOutcomeUtils.createOperationOutcome;
 
 @ControllerAdvice
 @RestController
 @Slf4j
 public class OperationOutcomeExceptionHandler extends ResponseEntityExceptionHandler {
 
-    private static final String CONTENT_TYPE = "Content type";
-    private static final String NOT_SUPPORTED = "not supported";
+    private static final Map<String, List<HttpMethod>> ALLOWED_METHODS = Map.of("/Bundle", List.of(POST));
```
[the remark this answers is not in the document]
Comment (dawidskainos) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/components/OperationOutcomeExceptionHandler.java:49]: Let's add ACS when we work on the story. What we have currently does not even have correct endpoint path.

**Review by bartek-sarul — APPROVED** [review]

**Pull request merged**